# IIJA All Projects 

In [1]:
# ! pip install nltk
# ! pip install textblob

In [2]:
import numpy as np
import pandas as pd
from siuba import *

from shared_utils import geography_utils
from dla_utils import _dla_utils

from calitp import to_snakecase

import utils

/opt/conda/lib/python3.10/site-packages/geopandas/_compat.py:112: UserWarning: The Shapely GEOS version (3.10.3-CAPI-1.16.1) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.


In [3]:
pd.set_option("display.max_columns", 100)
pd.set_option('display.max_colwidth', None)


## Read Data

In [4]:
GCS_FILE_PATH  = 'gs://calitp-analytics-data/data-analyses/dla/dla-iija'

In [5]:
proj = to_snakecase(pd.read_excel(f"{GCS_FILE_PATH}/CopyofFMIS_Projects_Universe_IIJA_Reporting_4.xls", 
                           # sheet_name='FMIS 5 Projects  ', header=[3]
                           sheet_name='IIJA',
                           # sheet_name='FMIS 5 Projects  ',
                           ))

In [6]:
proj.drop(columns =['unnamed:_0', 'unnamed:_13', 'unnamed:_14'], axis=1, inplace=True)

In [7]:
proj.sample(5)

fmis_transaction_date program_code        program_code_description  \
147             2022-09-07         Y003  PROJ TO REDUCE PM 2.5 EMI IIJA   
1037            2022-08-30         Y237         STBG 5K-49,999 POP IIJA   
412             2022-07-25         Y240  SURFAC TRNSP BLK GRTS-FLX IIJA   
664             2022-08-08         YS30    HIGHWAY SAFETY IMP PROG IIJA   
423             2022-04-04         Y233     STBG IIJA OFF-SYSTEM BRIDGE   

     project_number recipient_project_number  \
147         5085055              1021000034L   
1037        X093010              0217000009S   
412         5933143              0418000118L   
664         P020179              0300020624S   
423         5937058                04923198L   

                                                                                                                                              project_title  \
147   NOTTINGHAM LANE (FROM OLIVE TO HANSON), TAMWORTH CT, BUCKINGHAM CT, AND ALEXANDER AVE. INSTALL FIVE-FOOT SIDEWALK ADJACENT TO EXISTING CURB AND GUTTE   
1037  IN SISKIYOU COUNTY IN YREKA ON ROUTE 3 FROM 0.4 MILE NORTH OF LAURA LANE TO JUNIPER DRIVE AND ON ROUTE 263 FROM ROUTE 3 TO 1.0 MILE SOUTH OF LONG GUL   
412   IN CASTRO VALLEY: ON ANITA AVENUE BETWEEN CASTRO VALLEY BLVD. AND SOMERSET AVENUE CONSTRUCT SIDE WALKS,CURBS, GUTTERS, DRIVEWAYS, PEDESTRIAN RAMPS AN   
664   NEAR LAKE WILDWOOD FROM 0.4 MILE EAST OF MCGANNERY LANE TO YUBA/NEVADA COUNTY LINE; ALSO IN NEVADA COUNTY, FROM YUBA/NEVADA COUNTY LINE TO MOONEY FLA   
423                                                                         ALAMITOS RD. BRIDGE @ ALAMITOS CK (37C0159) BRIDGE REPLACEMENT/SEISMIC RET (TC)   

      county_code congressional_district  improvement_type  \
147         47.00           Cong Dist 16             28.00   
1037        93.00            Cong Dist 1              6.00   
412          1.00           Cong Dist 15             17.00   
664        115.00            Cong Dist 3             17.00   
423         85.00           Cong Dist 19             17.00   

                 improvement_type_description  obligations_amount  \
147   Facilities for Pedestrians and Bicycles          470,813.00   
1037        4R - Restoration & Rehabilitation       24,311,937.39   
412                  Construction Engineering            1,000.00   
664                  Construction Engineering          278,000.00   
423                  Construction Engineering        1,075,808.00   

     summary_recipient_defined_text_field_1_value  
147                                     L5085MCAG  
1037                                 S    NON-MPO  
412                                      L5933MTC  
664                                    S    SACOG  
423                                      L5937MTC

In [8]:
proj = proj.dropna(how='all') 

In [9]:
proj['summary_recipient_defined_text_field_1_value'] = proj['summary_recipient_defined_text_field_1_value'].fillna(value='None')

In [10]:
proj.head(1)

fmis_transaction_date program_code   program_code_description  \
0            2022-03-03         Y230  STBG-URBANIZED >200K IIJA   

  project_number recipient_project_number  \
0        5002174              0314000222L   

                                                             project_title  \
0  BELL AVE. FROM BOLLENBACHER AVE. TO ASTORIA ST ROAD REHABILITATION (TC)   

   county_code congressional_district  improvement_type  \
0        67.00            Cong Dist 6              4.00   

  improvement_type_description  obligations_amount  \
0       4R - No Added Capacity        2,841,397.00   

  summary_recipient_defined_text_field_1_value  
0                                   L5002SACOG

In [11]:
len(proj)

1241

In [12]:
## agency identifier column doesnt always have locode

## Function Work

In [13]:
# (proj[proj['summary_recipient_defined_text_field_1_value'].str.startswith('L')]).sample()

In [14]:
# (proj[proj['summary_recipient_defined_text_field_1_value'].str.startswith('L')]).sample()

In [15]:
#seeing if the space is the way to filter
# ((proj[~proj['summary_recipient_defined_text_field_1_value'].str.contains(" ")])>>count(_.summary_recipient_defined_text_field_1_value))

In [16]:
#check to make sure there are numbers
# ((proj[~proj['summary_recipient_defined_text_field_1_value'].str.contains(" ")])>>count(_.summary_recipient_defined_text_field_1_value)).iloc[1:20]

In [17]:
# locode_proj = ((proj[~proj['summary_recipient_defined_text_field_1_value'].str.contains(" ")]))
# locode_proj = locode_proj>>filter(_.summary_recipient_defined_text_field_1_value!='None')

In [18]:
# len(locode_proj)

In [19]:
# locode_proj = (utils.add_name_from_locode(locode_proj, 'summary_recipient_defined_text_field_1_value'))

In [20]:
# no_locode = ((proj[proj['summary_recipient_defined_text_field_1_value'].str.contains(" ")]))

In [21]:
# no_entry = proj>>filter(_['summary_recipient_defined_text_field_1_value']=='None')
    
#     #concat no locodes and those with no entry
# no_locode = pd.concat([no_locode, no_entry])

In [22]:
# no_locode.sample()

In [23]:
# no_locode = no_locode.rename(columns = {"recipient_name":"implementing_agency", "county_description":"county_name"})

In [24]:
# len(no_locode)

In [25]:
# pd.concat([locode_proj, no_locode])

## Function for adding Implemeting Agency Information

In [26]:
# function moved to utils

In [27]:
df = utils.identify_agency(proj, 'summary_recipient_defined_text_field_1_value')

In [28]:
df.sample()

fmis_transaction_date program_code    program_code_description  \
216            2022-09-07         Y400  CONGESTION MITIGATION IIJA   

    project_number recipient_project_number  \
216        5197043              0623000036L   

                                                                               project_title  \
216  INTERSECTION OF BETHEL AVE AND ALMOND AVE INSTALL NEW THREE PHASE TRAFFIC SIGNAL SYSTEM   

     county_code congressional_district  improvement_type  \
216        19.00           Cong Dist 21             15.00   

    improvement_type_description  obligations_amount  \
216      Preliminary Engineering           38,953.00   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
216                                    L5197FCOG                    5,197.00   

    implementing_agency  district    county_name  \
216              Sanger      6.00  Fresno County   

                                rtpa_name                             mpo_name  
216  Council of Fresno County Governments  Council Of Fresno County Goverments

In [29]:
df.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 1254 entries, 0 to 671
Data columns (total 18 columns):
 #   Column                                        Non-Null Count  Dtype         
---  ------                                        --------------  -----         
 0   fmis_transaction_date                         1254 non-null   datetime64[ns]
 1   program_code                                  1254 non-null   object        
 2   program_code_description                      1254 non-null   object        
 3   project_number                                1254 non-null   object        
 4   recipient_project_number                      1253 non-null   object        
 5   project_title                                 1254 non-null   object        
 6   county_code                                   1254 non-null   float64       
 7   congressional_district                        1254 non-null   object        
 8   improvement_type                              1254 non-null   float64

In [30]:
df = (utils.add_description(df, 'project_title'))

In [31]:
df>>filter(_.project_type=="Project")

fmis_transaction_date program_code      program_code_description  \
2              2022-08-26         Y400    CONGESTION MITIGATION IIJA   
3              2022-08-26         Y400    CONGESTION MITIGATION IIJA   
4              2022-07-15         Y400    CONGESTION MITIGATION IIJA   
5              2022-02-08         YS30  HIGHWAY SAFETY IMP PROG IIJA   
6              2022-02-08         YS30  HIGHWAY SAFETY IMP PROG IIJA   
..                    ...          ...                           ...   
662            2022-06-16         Y001    NATIONAL HIGHWAY PERF IIJA   
663            2022-06-16         Y001    NATIONAL HIGHWAY PERF IIJA   
664            2022-06-16         Y001    NATIONAL HIGHWAY PERF IIJA   
665            2022-06-16         Y001    NATIONAL HIGHWAY PERF IIJA   
671            2022-09-23         YR10   STATE REC TRAILS ADMIN IIJA   

    project_number recipient_project_number  \
2          5002178              0315000001L   
3          5002178              0315000001L   
4          5002189              0316000158L   
5          5002199              0317000138L   
6          5002199              0317000138L   
..             ...                      ...   
662        4053001              0716000031S   
663        4053001              0716000031S   
664        4053001              0716000031S   
665        4053001              0716000031S   
671        2023000                      NaN   

                                                                                                                                             project_title  \
2    SOUTH SACRAMENTO PARKWAY TRAIL- WEST. IN THE CITY OF SACRAMENTO, ADJACENT TO NORTHBOUND INTERSTATE 5, SOUTH OF POCKET ROAD/MEADOWVIEW ROAD AND FREEPO   
3    SOUTH SACRAMENTO PARKWAY TRAIL- WEST. IN THE CITY OF SACRAMENTO, ADJACENT TO NORTHBOUND INTERSTATE 5, SOUTH OF POCKET ROAD/MEADOWVIEW ROAD AND FREEPO   
4    ALONG UNUSED RAIL CORRIDOR NEAR FREEPORT BLVD FROM SUTTERVILLE ROAD TO SOUTH OF MEADOWVIEW/POCKET ROAD (4.5 MILES) RAILS TO TRAILS PROJECT ; NEW CLAS   
5    INTERSECTIONS OF 65TH ST EXPWY AT 21ST AVE, 65TH ST EXPWY AT FRUITRIDGE RD, ARDEN WAY AT HERITAGE LN, ARDEN WAY AT CHALLENGE WAY, HOWE AVE. AT SWARTH   
6    INTERSECTIONS OF 65TH ST EXPWY AT 21ST AVE, 65TH ST EXPWY AT FRUITRIDGE RD, ARDEN WAY AT HERITAGE LN, ARDEN WAY AT CHALLENGE WAY, HOWE AVE. AT SWARTH   
..                                                                                                                                                     ...   
662  IN LOS ANGELES COUNTY IN LOS ANGELES FROM 0.2 MILE SOUTH OF BURBANK BOULEVARD OVERCROSSING TO 0.3 MILE NORTH OF RINALDI STREET UNDERCROSSING PLANT VE   
663  IN LOS ANGELES COUNTY IN LOS ANGELES FROM 0.2 MILE SOUTH OF BURBANK BOULEVARD OVERCROSSING TO 0.3 MILE NORTH OF RINALDI STREET UNDERCROSSING PLANT VE   
664  IN LOS ANGELES COUNTY IN LOS ANGELES FROM 0.2 MILE SOUTH OF BURBANK BOULEVARD OVERCROSSING TO 0.3 MILE NORTH OF RINALDI STREET UNDERCROSSING PLANT VE   
665  IN LOS ANGELES COUNTY IN LOS ANGELES FROM 0.2 MILE SOUTH OF BURBANK BOULEVARD OVERCROSSING TO 0.3 MILE NORTH OF RINALDI STREET UNDERCROSSING PLANT VE   
671                                                                                  2023 REC TRAILS PROJECT: STATE PARKS INCURRED ADMINISTRATIVE EXPENSES   

     county_code congressional_district  improvement_type  \
2          67.00            Cong Dist 6             17.00   
3          67.00            Cong Dist 6             28.00   
4          67.00            Cong Dist 6             28.00   
5          67.00            Cong Dist 3             21.00   
6          67.00            Cong Dist 6             17.00   
..           ...                    ...               ...   
662        37.00           Cong Dist 29             44.00   
663        37.00           Cong Dist 30             17.00   
664        37.00           Cong Dist 30             42.00   
665        37.00           Cong Dist 30             44.00   
671   

In [32]:
df.sample()

fmis_transaction_date program_code    program_code_description  \
272            2022-05-05         Y001  NATIONAL HIGHWAY PERF IIJA   

    project_number recipient_project_number  \
272        P125020              1112000143S   

                                                                                                                                             project_title  \
272  IN SAN DIEGO COUNTY IN SAN DIEGO AND LEMON GROVE FROM 0.1 MILE SOUTH OF ELKELTON PLACE UNDERCROSSING TO ROUTE 94/125 SEPARATION. CONSTRUCT MAINTENANC   

     county_code congressional_district  improvement_type  \
272        73.00           Cong Dist 53             17.00   

    improvement_type_description  obligations_amount  \
272     Construction Engineering        1,234,100.00   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
272                                  S    SANDAG                         NaN   

    implementing_agency  district       county_name rtpa_name mpo_name  \
272          California     11.00  San Diego County       NaN      NaN   

    project_method project_type  
272      Construct      Project

In [33]:
df.loc[df.county_name == "Statewide County", 'county_name'] = "Statewide"

## New Title Function

In [34]:
def add_new_title(df, first_col_method, second_col_type, third_col_name, alt_col_name):
    """
    Function to add new title. 
    Expected output example: "New Bike Lane in Eureka"
    """
    def return_name(df):
        
        if (df[third_col_name] == "California") & (df[alt_col_name] == "Statewide"):
            return (df[first_col_method] + " " + df[second_col_type] +" " + df[alt_col_name])
        
        elif (df[third_col_name] == "California"):
            return (df[first_col_method] + " " + df[second_col_type] + " in " + df[alt_col_name])
        
        elif (df[third_col_name] != "California"):
            return (df[first_col_method] + " " + df[second_col_type] + " in " + df[third_col_name])
        
        # elif (df[third_col_name] == "Metropolitan Transportation Commission"):
        #     return (df[first_col_method] + " " + df[second_col_type] + " in The " + df[third_col_name])

        return df

    df['project_name_new'] = df.apply(return_name, axis = 1)
    
    return df

In [35]:
df = add_new_title(df, "project_method", "project_type", "implementing_agency", "county_name")

In [36]:
df.tail()


fmis_transaction_date program_code      program_code_description  \
667            2022-09-09         YS30  HIGHWAY SAFETY IMP PROG IIJA   
668            2022-09-09         YS30  HIGHWAY SAFETY IMP PROG IIJA   
669            2022-09-09         YS30  HIGHWAY SAFETY IMP PROG IIJA   
670            2022-09-09         YS32   SEC 164 PENALTIES HSIP IIJA   
671            2022-09-23         YR10   STATE REC TRAILS ADMIN IIJA   

    project_number recipient_project_number  \
667        5801081              0416000125S   
668        5801081              0416000125S   
669        5801081              0416000125S   
670        5801081              0416000125S   
671        2023000                      NaN   

                                                                                                                     project_title  \
667  ALAMEDA COUNTY NEAR LIVERMORE FROM GRANT LINE ROAD UNDERCROSSING TO 0.1 MILE EAST OF NORTH FLYNN ROAD INSTALL LIGHTING SYSTEM   
668  ALAMEDA COUNTY NEAR LIVERMORE FROM GRANT LINE ROAD UNDERCROSSING TO 0.1 MILE EAST OF NORTH FLYNN ROAD INSTALL LIGHTING SYSTEM   
669  ALAMEDA COUNTY NEAR LIVERMORE FROM GRANT LINE ROAD UNDERCROSSING TO 0.1 MILE EAST OF NORTH FLYNN ROAD INSTALL LIGHTING SYSTEM   
670  ALAMEDA COUNTY NEAR LIVERMORE FROM GRANT LINE ROAD UNDERCROSSING TO 0.1 MILE EAST OF NORTH FLYNN ROAD INSTALL LIGHTING SYSTEM   
671                                                          2023 REC TRAILS PROJECT: STATE PARKS INCURRED ADMINISTRATIVE EXPENSES   

     county_code congressional_district  improvement_type  \
667         1.00           Cong Dist 15             21.00   
668         1.00           Cong Dist 15             22.00   
669         1.00           Cong Dist 15             42.00   
670         1.00           Cong Dist 15             17.00   
671       999.00            Cong Dist 0             44.00   

    improvement_type_description  obligations_amount  \
667                       Safety        4,165,800.00   
668            Rail/Hwy Crossing           18,900.00   
669                     Training            3,000.00   
670     Construction Engineering          288,000.00   
671                        Other          241,760.00   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
667                                     S    MTC                         NaN   
668                                     S    MTC                         NaN   
669                                     S    MTC                         NaN   
670                                     S    MTC                         NaN   
671                                         None                         NaN   

    implementing_agency  district     county_name rtpa_name mpo_name  \
667          California      4.00  Alameda County       NaN      NaN   
668          California      4.00  Alameda County       NaN      NaN   
669          California      4.00  Alameda County       NaN      NaN   
670          California      4.00  Alameda County       NaN      NaN   
671          California       NaN       Statewide       NaN      NaN   

    project_method project_type                    project_name_new  
667        Install     Lighting  Install Lighting in Alameda County  
668        Install     Lighting  Install Lighting in Alameda County  
669        Install     Lighting  Install Lighting in Alameda County  
670        Install     Lighting  Install Lighting in Alameda County  
671                     Project                   Project Statewide

In [37]:
df>>filter(_.implementing_agency==("Metropolitan Transportation Commission"))>>select(_.project_name_new)

project_name_new
518             Bike Share Program in Metropolitan Transportation Commission
519                        Project in Metropolitan Transportation Commission
520                        Signage in Metropolitan Transportation Commission
521                        Project in Metropolitan Transportation Commission
522                        Project in Metropolitan Transportation Commission
523                        Project in Metropolitan Transportation Commission
524   Regional Planning Activities in Metropolitan Transportation Commission
525   Regional Planning Activities in Metropolitan Transportation Commission
526   Regional Planning Activities in Metropolitan Transportation Commission
527   Regional Planning Activities in Metropolitan Transportation Commission
528   Regional Planning Activities in Metropolitan Transportation Commission
529   Regional Planning Activities in Metropolitan Transportation Commission
530   Regional Planning Activities in Metropolitan Transportation Commission
531   Regional Planning Activities in Metropolitan Transportation Commission
532   Regional Planning Activities in Metropolitan Transportation Commission
533   Regional Planning Activities in Metropolitan Transportation Commission
534   Regional Planning Activities in Metropolitan Transportation Commission
535   Regional Planning Activities in Metropolitan Transportation Commission
536              Implement Project in Metropolitan Transportation Commission
537              Implement Project in Metropolitan Transportation Commission
538              Implement Project in Metropolitan Transportation Commission

### Export progress

In [38]:
#test_df = utils.title_column_names(df)

In [39]:
#test_df.to_csv(f"{GCS_FILE_PATH}/FMIS_projects_wip_all.csv")

## Use Alt Column

In [59]:
df.sample()

fmis_transaction_date program_code    program_code_description  \
344            2022-06-16         Y001  NATIONAL HIGHWAY PERF IIJA   

    project_number recipient_project_number  \
344        Q101395              0520000169S   

                                                                                                                                              project_title  \
344  SANTA BARBARA COUNTY IN AND NEAR CARPINTERIA AND AT SUMMERLAND FROM 0.9 MILE SOUTH OF SOUTH PADARO LANE UNDERCROSSING TO 0.6 MILE NORTH OF PADARO LANE   

     county_code congressional_district  improvement_type  \
344        83.00           Cong Dist 24             42.00   

    improvement_type_description  obligations_amount  \
344                     Training           19,200.00   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
344                                   S    SBCAG                         NaN   

    implementing_agency  district           county_name rtpa_name mpo_name  \
344          California      5.00  Santa Barbara County       NaN      NaN   

    project_method project_type                  project_name_new  
344                     Project   Project in Santa Barbara County

In [60]:
# using those with no match

no_match = (df>>filter(_.project_type==('Project'))>>select(_.project_number, _.recipient_project_number, _.project_title, _.improvement_type_description, _.implementing_agency, _.county_name,
                                                            _.project_method, _.project_type, _.project_name_new))

In [62]:
no_match.improvement_type_description.value_counts()

Construction Engineering                               180
Safety                                                  74
Training                                                69
4R - Restoration & Rehabilitation                       52
Planning                                                44
Preliminary Engineering                                 38
4R - Maintenance  Resurfacing                           19
Facilities for Pedestrians and Bicycles                 18
Other                                                   16
Bridge Replacement - No Added Capacity                  14
Right of Way                                            13
4R - No Added Capacity                                  12
Bridge Rehabilitation - No Added Capacity                6
4R - Added Capacity                                      6
Mitigation of Water Pollution due to Highway Runoff      4
New  Construction Roadway                                4
Bridge Replacement - Added Capacity                     

In [43]:
no_match>>filter(_.improvement_type_description=="Preliminary Engineering")

project_title  \
135  BARTON ROAD OVERHEAD AT UNION PACIFIC RAILROAD, 0.25 MILE WEST OF I-215, BR. NO. 54C-0379 REPLACE OVERHEAD WITH TWO-LANE ROAD ACROSS DISCONTINUED RAI   
149  NORTH SIDE OF CHILDS AVE (FROM GOLDEN VALLEY HIGH SCHOOL TO COFFEE ST. (WEAVER MIDDLE SCHOOL)). CONSTRUCT A CLASS I (FULLY-SEPARATED). "SAFE ROUTES T   
158  PM00222/PSTUDY002 ON VARIOUS BRIDGE LOCATIONS THROUGHOUT THE CITY OF CORONA.  SEE COMMENTS FOR THE LIST OF BRIDGES. DEVELOP A BPMP PLAN FOR THE CITY-   
231                                                                                  VARIOUS LOCATION IN SUNNYVALE PROVIDE ADVANCE DILEMMA ZONE DETECTION.   
256                                                                                           US-50 AT EMPIRE RANCH ROAD CONSTRUCT 4 LANE INTERCHANGE (TC)   
294  PROJECT SPANS SACRAMENTO RIVER FROM 2ND/D STREET IN WEST SACRAMENTO, UP TO THE TOP DECK OF THE I STREET BRIDGE TO THE JIBBOOM/I STREET VIADUCT AND IN   
298  CITY OF MISSION VIEJO - THE PROJECT IS LOCATED ON THE WESTERN EDGE OF MISSION VIEJO. THE MISSION VIEJO N/S CORRIDOR TRAIL IS BOUNDED BY LOS ALISOS BL   
303  E/B SANTA MARGARITA PARKWAY OVER ARROYO TRABUCO (BRIDGE #55C0520R) AND ANTONIO PARKWAY OVER TIJERAS CREEK (BRIDGE #55C0605), BPMP PM00035 BRIDGE PREV   
307                                                               REQUA ROAD OVER HUNTER CREEK, 0.1 MI. WEST OF SR-101 REPLACE EXISTING 2-LANE BRIDGE (TC)   
383  IN EASTERN EI DORADO COUNTY, IN THE TAHOE BASIN, IN THE CITY OF SOUTH LAKE TAHOE AT EAST END OF SOUTH AVENUE, OVER THE UPPER TRUCKEE RIVER, THEN NORT   
388  4 AT TERNERS DRIVE @ DRIVEWAY ENTRANCES TO MULTI-FAMILY HOUSING CLOSEST TO DONAHUE, 4 AT TERNERS DRIVE @ TERRACE WAY, 4 AT TERNERS DRIVE @ TERRACE DR   
424  ALDER CROFT HEIGHTS ROAD OVER HOOKER CREEK, 2.12 MILES SOUTH OF LOS GATOS SC ROAD.   BRIDGE NO. 37C0506 REPLACE EXISTING ONE LANE BRIDGE WITH A NEW T   
426  ALDERCROFT HEIGHTS ROAD OVER LOS GATOS CREEK, 2.12 MILES S OF LOS GATOS CREEL ROAD. BRIDGE 37C0507 REPLACE EXISTING ONE LANE BRIDGE WITH A NEW 2 LANE   
428  ALDERCROFT HEIGHTS ROAD OVER LOS GATOS CREEK 2.38 MILES SOUTH OF LOS GATOS SC ROAD. BRIDGE 37C0508 REPLACE EXISTING ONE-LANE BRIDGE WITH A NEW 2-LANE   
448                                                                   BRIDGE 41C0032 ON AVENUE 25 OVER ASH SLOUGH REPLACE 2 LANE BRIDGE WITH 2 LANE BRIDGE   
452                                                   BRIDGE 41C0099 COUNTY ROAD 5 1/2 OVER CHOWCHILLA CREEK REPLACE 2 LANE BRIDGE WITH 2 LANE BRIDGE (TC)   
454                                                           BRIDGE 42C0175 ON MANNING AVENUE OVER TRAVERS CREEK REPLACE 2 LANE BRIDGE WITH 4 LANE BRIDGE   
460                                                   BRIDGE 42C0270 ON MILLERTON ROAD OVER LITTLE DRY CREEK REPLACE 2 LANE BRIDGE WITH 2 LANE BRIDGE (TC)   
472                                    BRIDGE 46C0196 ON M375A MINERAL KING ROAD OVER EAST FORK KAWEAH RIVER REPLACE 2 LANE BRIDGE WITH 2 LANE BRIDGE (TC)   
473                                                 BRIDGE 46C0195 ON ROAD M348 OVER SOUTH FORK KAWEAH RIVER REPLACE 1 LANE BRIDGE WITH 1 LANE BRIDGE (TC)   
479  GENESEO RD LOW WATER CROSSING OVER HUER HUERO CREEK - BR. NO. 00L0040 (49C-0431 NBI BRIDGE NEARBY)RELATED TO 5949(180) FOR LONG-TERM POST-CONSTRUCTIO   
516  STATE ROUTE 55 FROM INTERSTATE 5 TO STATE ROUTE 91 ADD A MIXED-FLOW LANE IN EACH DIRECTION AND FIX CHOKEPOINTS FROM I-5 TO SR-22 AND OTHER IMPROVEMEN   
517  STATE ROUTE 55 FROM INTERSTATE 5 TO STATE ROUTE 91 ADD A MIXED-FLOW LANE IN EACH DIRECTION AND FIX CHOKEPOINTS FROM I-5 TO SR-22 AND OTHER IMPROVEMEN   
522  SF BAY AREA: VARIOUS BRIDGE CORRIDORS AND CORRIDOR APPROACHES CONDUCT PRELIMINARY ENGINEERING AND PLANNING STUDIES TO ADVANCE BAY BRIDGE FORWARD PROJ   
523  SF BAY AREA: VARIOUS BRIDGE CORRIDORS AND CORRIDOR APPROACHES CONDUCT PRELIMINARY ENGINEERING AND PLANNING STUDIES TO ADVANCE BAY BRIDGE FORWARD PROJ   
545  ALONG THE I-880 CORRIDOR SF 

In [44]:
((utils.get_list_of_words((no_match>>filter(_.improvement_type_description=="4R - Restoration & Rehabilitation")), "project_title")).value_counts()).iloc[0:20]

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /home/jovyan/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


mile             40
route            29
county           29
san              21
near             19
south            18
road             16
north            14
east             13
street           13
construct        13
undercrossing    13
05               13
various          11
diego            11
02               11
locations        11
los              11
separation       11
west             10
dtype: int64

In [45]:
no_match>>filter(_.project_title.str.contains("BICYC"), _.improvement_type_description!="Facilities for Pedestrians and Bicycles")

Empty DataFrame
Columns: [project_title, improvement_type_description, implementing_agency, county_name, project_method, project_type, project_name_new]
Index: []

In [46]:
(no_match>>filter(_.improvement_type_description=="4R - Restoration & Rehabilitation"))

project_title  \
79   OAKPORT ST FROM EDGEWATER TO I-880 FREEWAY ENTRANCE; OAKLAND AVE FROM ORANGE STREET TO MACARTHUR; MONTEREY BLVD FROM MAIDEN LN TO BENNET PL IN OAKLAN   
183  EAST DUNNE AVENUE BETWEEN LOWER THOMAS GRADE AND THE EASTERN CITY LIMIT. FEDERAL PARTICIPATING LOCATION SEGMENTS: 1) LOWER THOMAS GRADE TO FLAMING OA   
232  IN EAST SUNNYVALE SENSE OF PLACE PLAN AREA: ON DUANE AVENUE FROM FAIR OAKS AVENUE TO LAWRENCE EXPRESSWAY, STEWART DRIVE FROM WOLFE ROAD TO DUANE AVEN   
247  BELMONT: CHULA VISTA FROM ALAMEDA DE LAS PULGAS TO RALSTON AVE, 6TH AVENUE FROM RALSTON AVENUE TO HILL STREET, 6TH AVENUE FROM EMMETT AVENUE TO HARBO   
281  COLEMAN CREEK PATH @ SNYDER LANE, HINEBAUGH CREEK PATH @ COUNTRY CLUB DRIVE, HINEBAUGH CREEK PATH @ STATE FARM DRIVE, HINEBAUGH CREEK PATH @ COMMERCE   
299  GOLDEN VALLEY PARKWAY (TOWNE CENTRE DRIVE AND LATHROP ROAD), RIVER ISLANDS PARKWAY (GOLDEN VALLEY PARKWAY LATHROP FIRE STATION 34), LATHROP ROAD. ROA   
579   IN AND NEAR THE CITY OF MODESTO ON SR 132 NEAR NORTH DAKOTA AVE TO NEAR 6TH ST AND ON SR 99 FROM 0.5 M SOUTH OF TUOLUMNE RIVER BRIDGE TO 0.1 M NORTH   
1                                                                   ALAMEDA COUNTY AT SUNOL AND IN PLEASANTON AND DUBLIN PRECAST CONCRETE PAVEMENT AND HMA   
4    IN SAN DIEGO COUNTY IN CHULA VISTA, FROM 0.3 MILE SOUTH OF MAIN STREET UNDERCROSSING TO PALOMAR STREET OVERCROSSING CONVERT POTABLE IRRIGATION SYSTEM   
6    IN SAN DIEGO COUNTY IN CHULA VISTA, FROM 0.3 MILE SOUTH OF MAIN STREET UNDERCROSSING TO PALOMAR STREET OVERCROSSING CONVERT POTABLE IRRIGATION SYSTEM   
12   ALAMEDA COUNTY NEAR OAKLAND AT VARIOUS LOCATIONS FROM 0.1 MILE NORTH OF 7TH STREET UNDERCROSSING TO 0.5 MILE NORTH OF 16TH AVE OVERCROSSING. CRASH CU   
29                                                      IN ALAMEDA AND SOLANO COUNTIES AT VARIOUS LOCATION INSTALL HIGHWAY PLANTING AND IRRIGATION SYSTEMS   
39   SANTA CLARA AND ALAMEDA COUNTIES ON ROUTE 880 AND 80 AT VARIOUS LOCATIONS FROM ROUTE 880/237 CONNECTOR SEPARATION TO 98TH AVE OVER CROSSING AND ON RO   
50                                     SLO CO ABOUT 19 MI N OF SAN SIMEON FR 0.7 MI TO 1.0 MI N OF RAGGED PT VIADUCT CONSTRUCT SOLDIER PILE RETAINING WALL   
71   LOS ANGELES COUNTY IN SANTA CLARITA FROM 0.9 MILE NORTH OF PLACERITA CANYON ROAD UNDERCROSSING TO 0.2 MILE SOUTH OF GOLDEN VALLEY ROAD OVERCROSSING P   
79   SANTA CLARA COUNTY IN AND NEAR LOS GATOS AND SAN JOSE FROM HEBARD ROAD TO ROUTE 17/280 SEPERATION COLD PLAN AND PAVE AC, REMOVE AND INSTALL CULVERTS,   
88   NEVADA COUNTY NEAR OMEGA AT VARIOUS LOCATIONS FROM 0.2 MILE WEST OF CONSERVATION ROAD (WASHINGTON RIDGE ROAD) TO 1.2 MILES WEST OF ZIEBRIGHT ROAD. RO   
141   SACRAMENTO COUNTY IN SACRAMENTO FROM 0.1 MILE SOUTH OF B STREET UNDERPASS TO 0.4MILE NORTH OF ARDEN WAY UNDERCROSSING. CONSTRUCT MAINTENANCE VEHICLE   
218                                              SAN MATEO COUNTY, IN SOUTH SAN FRANCISCO AT 0.1 MILE NORTH OF FRANCISCO DRIVE CONSTRUCT SOLDIER PILE WALL   
257  IN THE CITY OF LOS ANGELES, FROM TOPANGA CANYON BOULEVARD TO SEPULVEDA BOULEVARD. CONSTRUCT AND INSTALL STORMWATER  TREATMENT BEST MANAGEMENT PRACTIC   
260  IN THE CITY OF LOS ANGELES, FROM TOPANGA CANYON BOULEVARD TO SEPULVEDA BOULEVARD. CONSTRUCT AND INSTALL STORMWATER  TREATMENT BEST MANAGEMENT PRACTIC   
263  IN THE CITY OF LOS ANGELES, FROM TOPANGA CANYON BOULEVARD TO SEPULVEDA BOULEVARD. CONSTRUCT AND INSTALL STORMWATER  TREATMENT BEST MANAGEMENT PRACTIC   
271  IN SAN DIEGO COUNTY IN SAN DIEGO AND LEMON GROVE FROM 0.1 MILE SOUTH OF ELKELTON PLACE UNDERCROSSING TO ROUTE 94/125 SEPARATION. CONSTRUCT MAINTENANC   
312   ORANGE COUNTY AT VARIOUS LOCATIONS FROM 0.1 MILE NORTH OF ROUTE 133 TO 0.5 MILESOUTH OF ROUTE 91 PAVEMENT REPAIR, GROUND WATER EXCLUSION AND TRAFFIC   
337                                 IN DEL NORTE COUNTY NEAR CRESCENT CITY AT 0.2 MILE NORTH OF CUSHING CREEK VIADUCT R RESTORE ROADWAY / LANDSLIDE REPAIR   
345   SANTA BARBARA AND SAN LUIS 

### Function for no_matches

In [154]:
def update_no_matched(df, flag_col, desc_col, name_col):
    """
    function to itreate over projects that did not match the first time
    using an existing project's short description of project type. 
    """
    
    def return_project_type(df):
        
        if (df[flag_col] == "Project") & (df[desc_col] == "Bridge Rehabilitation") | (df[desc_col] =="Bridge Rehabilitation - No Added Capacity") | (df[desc_col] =="Bridge Rehabilitation - Added Capacity"):
            return ("Bridge Rehabilitation in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "Facilities for Pedestrians and Bicycles"):
            return (df[desc_col] + " in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "Safety"):
            return (df[desc_col] + " Improvement Projects in " + df[name_col])
        
        # elif (df[flag_col] == "Project") & (df[flag_col] == "Planning") & (df[title_col].str.contains("REGION CONSOLIDATED PLANNING GRANT")):
        #     return ("Regional Planning Grant in " + df[name_col])
            
        elif (df[flag_col] == "Project") & (df[desc_col] == "Planning "):
            return "Project Planning in " + df[name_col]
            
        elif (df[flag_col] == "Project") & (df[desc_col] == "Preliminary Engineering"):
            return (df[desc_col] + " Projects in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "Construction Engineering"):
            return (df[desc_col] + " Projects in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "4R - Restoration & Rehabilitation"):
            return ("Road Restoration & Rehabilitation in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "4R - Maintenance  Resurfacing"):
            return ("Maintenance Resurfacing in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "Bridge Replacement - Added Capacity") | (df[desc_col] == "Bridge Replacement - No Added Capacity") | (df[desc_col] == "Bridge New Construction") | (df[desc_col] == "Special Bridge"):
            return ("Bridge Replacement in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "Mitigation of Water Pollution due to Highway Runoff"):
            return (df[desc_col] + " in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "4R - Added Capacity"):
            return ("Added Roadway Capacity in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "4R - No Added Capacity"):
            return ("Road Construction in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "New  Construction Roadway"):
            return ("New Construction Roadway in " + df[name_col])
        
        elif (df[flag_col] == "Project") & (df[desc_col] == "Traffic Management/Engineering - HOV"):
            return ("Traffic Management Project in " + df[name_col])
        
        # elif (df[flag_col] == "Project") & (df[desc_col] == "Training"):
        #     return (df[desc_col] + " for ___ in " + df[name_col])
        
        else:
            return "" #(df[desc_col] + " in " + df[name_col])

        return df


    df['project_name_new2'] = df.apply(return_project_type, axis = 1)
    
    #df.apply(func, axis=1)
    
    return df

In [155]:
no_match_title = (update_no_matched(no_match, 'project_type', 'improvement_type_description', 'implementing_agency'))

In [156]:
no_match_title.sample()

project_number recipient_project_number  \
655        2801153              0416000024S   

                                                                                                                                       project_title  \
655  SANTA CLARA COUNTY IN AND NEAR SAN JOSE FROM S280-N101/101 SEPARATION TO ROUTE 17/280 SEPERATION MINOR CONCRETE, MGS RAILING, HMA AND HMA DIKE.   

    improvement_type_description implementing_agency         county_name  \
655     Construction Engineering          California  Santa Clara County   

    project_method project_type                project_name_new  \
655                     Project   Project in Santa Clara County   

                                   project_name_new2  
655  Construction Engineering Projects in California

In [50]:
# remaining = (no_match_title>>filter(_.project_name_new2=='')).improvement_type_description.value_counts()
# remaining

### Remaining 
| Description | Count |
| --- | --- |
|Training | 69 |
| Other  |  16 |
| Right of Way | 13 |
|  Safety and Education for Peds/Bicyclist | 1 |
| Bridge Inspection and Bridge Related Training | 1 |
| Rail/Hwy Crossing    |  1 |
| Utilities    | 1 |
| Landscaping and Other Scenic Beautification   |    1 |
| Research   |  1 |


In [58]:
df>>filter(_.project_title.str.contains("CHOKEPOINT RELIEF"))

fmis_transaction_date program_code        program_code_description  \
576            2022-09-22         Y460  NATIONAL HWY FREIGHT PROG IIJA   
577            2022-09-22         Y460  NATIONAL HWY FREIGHT PROG IIJA   
578            2022-09-22         Y460  NATIONAL HWY FREIGHT PROG IIJA   

    project_number recipient_project_number  \
576        P057073              0715000076S   
577        P057073              0715000076S   
578        P057073              0715000076S   

                                                                                                                                             project_title  \
576  IN LOS ANGELES COUNTY. ROUTE 57/60 CONFLUENCE CHOKEPOINT RELIEF PROGRAM. RECONSTRUCT GRAND AVENUE OVERCROSSING. RECONSTRUCT NORTHBOUND SR-57 CONNECTO   
577  IN LOS ANGELES COUNTY. ROUTE 57/60 CONFLUENCE CHOKEPOINT RELIEF PROGRAM. RECONSTRUCT GRAND AVENUE OVERCROSSING. RECONSTRUCT NORTHBOUND SR-57 CONNECTO   
578  IN LOS ANGELES COUNTY. ROUTE 57/60 CONFLUENCE CHOKEPOINT RELIEF PROGRAM. RECONSTRUCT GRAND AVENUE OVERCROSSING. RECONSTRUCT NORTHBOUND SR-57 CONNECTO   

     county_code congressional_district  improvement_type  \
576        37.00           Cong Dist 39             11.00   
577        37.00           Cong Dist 39             17.00   
578        37.00           Cong Dist 39             42.00   

               improvement_type_description  obligations_amount  \
576  Bridge Replacement - No Added Capacity       72,969,500.00   
577                Construction Engineering        1,000,000.00   
578                                Training           30,500.00   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
576                                    L6065SCAG                    6,065.00   
577                                    L6065SCAG                    6,065.00   
578                                    L6065SCAG                    6,065.00   

                                          implementing_agency  district  \
576  Los Angeles County Metropolitan Transportation Authority      7.00   
577  Los Angeles County Metropolitan Transportation Authority      7.00   
578  Los Angeles County Metropolitan Transportation Authority      7.00   

            county_name                                             rtpa_name  \
576  Los Angeles County  Los Angeles County Metropolitan Transportation Auth.   
577  Los Angeles County  Los Angeles County Metropolitan Transportation Auth.   
578  Los Angeles County  Los Angeles County Metropolitan Transportation Auth.   

                                           mpo_name project_method  \
576  Southern California Association Of Governments      Construct   
577  Southern California Association Of Governments      Construct   
578  Southern California Association Of Governments      Construct   

    project_type  \
576      Project   
577      Project   
578      Project   

                                                                  project_name_new  
576  Construct Project in Los Angeles County Metropolitan Transportation Authority  
577  Construct Project in Los Angeles County Metropolitan Transportation Authority  
578  Construct Project in Los Angeles County Metropolitan Transportation Authority

In [52]:
no_match_title>>filter(_.improvement_type_description=='Training')

project_title  \
578  IN LOS ANGELES COUNTY. ROUTE 57/60 CONFLUENCE CHOKEPOINT RELIEF PROGRAM. RECONSTRUCT GRAND AVENUE OVERCROSSING. RECONSTRUCT NORTHBOUND SR-57 CONNECTO   
580   IN AND NEAR THE CITY OF MODESTO ON SR 132 NEAR NORTH DAKOTA AVE TO NEAR 6TH ST AND ON SR 99 FROM 0.5 M SOUTH OF TUOLUMNE RIVER BRIDGE TO 0.1 M NORTH   
3                                                                   ALAMEDA COUNTY AT SUNOL AND IN PLEASANTON AND DUBLIN PRECAST CONCRETE PAVEMENT AND HMA   
11                              ALAMEDA COUNTY IN SAN LEANDRO AT WASHINGTON AVENUE OVERCROSSING ROADWAY EXCAVATION, HMA, AND MODIFYING ELECTRICAL SYSTEMS.   
18                                                                     ALAMEDA AND CONTRA COSTA COUNTIES AT VARIOUS LOCATIONS CROSSWALK SAFETY ENHANCEMENT   
..                                                                                                                                                     ...   
650  IN RIVERSIDE COUNTY NEAR MORENO VALLEY ON INTERSTATE 215 FROM 0.2 MILE NORTH OF ALESSANDRO BOULEVARD OVERCROSSING TO 0.2 MILE SOUTH OF EUCALYPTUS AVE   
653        SANTA CLARA COUNTY IN AND NEAR SAN JOSE FROM S280-N101/101 SEPARATION TO ROUTE 17/280 SEPERATION MINOR CONCRETE, MGS RAILING, HMA AND HMA DIKE.   
656        SANTA CLARA COUNTY IN AND NEAR SAN JOSE FROM S280-N101/101 SEPARATION TO ROUTE 17/280 SEPERATION MINOR CONCRETE, MGS RAILING, HMA AND HMA DIKE.   
661  IN LOS ANGELES COUNTY IN LOS ANGELES FROM 0.2 MILE SOUTH OF BURBANK BOULEVARD OVERCROSSING TO 0.3 MILE NORTH OF RINALDI STREET UNDERCROSSING PLANT VE   
664  IN LOS ANGELES COUNTY IN LOS ANGELES FROM 0.2 MILE SOUTH OF BURBANK BOULEVARD OVERCROSSING TO 0.3 MILE NORTH OF RINALDI STREET UNDERCROSSING PLANT VE   

    improvement_type_description  \
578                     Training   
580                     Training   
3                       Training   
11                      Training   
18                      Training   
..                           ...   
650                     Training   
653                     Training   
656                     Training   
661                     Training   
664                     Training   

                                          implementing_agency  \
578  Los Angeles County Metropolitan Transportation Authority   
580                                                   Modesto   
3                                                  California   
11                                                 California   
18                                                 California   
..                                                        ...   
650                                                California   
653                                                California   
656                                                California   
661                                                California   
664                                                California   

            county_name project_method project_type  \
578  Los Angeles County      Construct      Project   
580   Stanislaus County                     Project   
3        Alameda County           Pave      Project   
11       Alameda County                     Project   
18            Statewide                     Project   
..                  ...            ...          ...   
650    Riverside County                     Project   
653  Santa Clara County                     Project   
656  Santa Clara County                     Project   
661  Los Angeles County                     Project   
664  Los Angeles County                     Project   

                                                                  project_name_new  \
578  Construct Project in Los Angeles County Metropolitan Transportation Authority   
580                                                             Project in Modesto   
3                                                   Pave Project in Alameda County   
11                       

In [53]:
((utils.get_list_of_words((no_match_title>>filter(_.improvement_type_description=='Training')), "project_title")).value_counts()).iloc[0:40]

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /home/jovyan/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


mile             64
county           50
route            35
near             31
north            25
road             25
south            25
los              23
san              22
west             22
02               20
angeles          19
overcrossing     18
avenue           16
boulevard        14
construct        13
santa            13
hma              13
locations        12
01               12
east             12
various          12
bridge           11
05               11
separation       10
undercrossing    10
miles            10
04                9
canyon            8
03                8
orange            7
city              7
street            6
counties          6
systems           6
install           6
replace           6
drainage          6
upgrade           6
safety            5
dtype: int64

In [54]:
# df>>filter(_.improvement_type_description=='Training')>>select(_.program_code_description, _.project_title, _.improvement_type_description,
#                                                               _.project_method, _.project_type, _.project_name_new)

In [55]:
## potential duplicates. different "descriptions 

df>>filter(_.project_title.str.contains("ERN COUNTY IN THE CITY OF BAKERSFIELD ON ROUTE 99 FROM 0.3 MILE SOUTH OF BELLE TERRACE OVERCROSSING TO 0.1 MILE NORTH OF ROUTE 58"))>>select(_.program_code_description, _.project_title, _.improvement_type_description,
                                                              _.project_method, _.project_type, _.project_name_new)

program_code_description  \
398  NATIONAL HIGHWAY PERF IIJA   
399  NATIONAL HIGHWAY PERF IIJA   
400  NATIONAL HIGHWAY PERF IIJA   
401  NATIONAL HIGHWAY PERF IIJA   
402  NATIONAL HIGHWAY PERF IIJA   
403  NATIONAL HIGHWAY PERF IIJA   

                                                                                                                                            project_title  \
398  KERN COUNTY IN THE CITY OF BAKERSFIELD ON ROUTE 99 FROM 0.3 MILE SOUTH OF BELLE TERRACE OVERCROSSING TO 0.1 MILE NORTH OF ROUTE 58; ALSO ON ROUTE 58   
399  KERN COUNTY IN THE CITY OF BAKERSFIELD ON ROUTE 99 FROM 0.3 MILE SOUTH OF BELLE TERRACE OVERCROSSING TO 0.1 MILE NORTH OF ROUTE 58; ALSO ON ROUTE 58   
400  KERN COUNTY IN THE CITY OF BAKERSFIELD ON ROUTE 99 FROM 0.3 MILE SOUTH OF BELLE TERRACE OVERCROSSING TO 0.1 MILE NORTH OF ROUTE 58; ALSO ON ROUTE 58   
401  KERN COUNTY IN THE CITY OF BAKERSFIELD ON ROUTE 99 FROM 0.3 MILE SOUTH OF BELLE TERRACE OVERCROSSING TO 0.1 MILE NORTH OF ROUTE 58; ALSO ON ROUTE 58   
402  KERN COUNTY IN THE CITY OF BAKERSFIELD ON ROUTE 99 FROM 0.3 MILE SOUTH OF BELLE TERRACE OVERCROSSING TO 0.1 MILE NORTH OF ROUTE 58; ALSO ON ROUTE 58   
403  KERN COUNTY IN THE CITY OF BAKERSFIELD ON ROUTE 99 FROM 0.3 MILE SOUTH OF BELLE TERRACE OVERCROSSING TO 0.1 MILE NORTH OF ROUTE 58; ALSO ON ROUTE 58   

               improvement_type_description project_method project_type  \
398                  4R - No Added Capacity                     Project   
399                  4R - No Added Capacity                     Project   
400  Bridge Replacement - No Added Capacity                     Project   
401  Bridge Replacement - No Added Capacity                     Project   
402                                Training                     Project   
403                                Training                     Project   

            project_name_new  
398   Project in Kern County  
399   Project in Kern County  
400   Project in Kern County  
401   Project in Kern County  
402   Project in Kern County  
403   Project in Kern County

In [56]:
no_match_title >>filter(_.project_title.str.contains('BRIDGE'))>>select(_.project_title, _.improvement_type_description,
                                                              _.project_method, _.project_type, _.project_name_new, _.project_name_new2)>>filter(_.project_name_new2=="")

Empty DataFrame
Columns: [project_title, improvement_type_description, project_method, project_type, project_name_new, project_name_new2]
Index: []

### Try with one element of project

- projects listed can have multiple entires bases on the phases they have. currently, if a project has multiple entries, the `improvement_type_description` has different description phrases, leading the title function to create a title that does not accuratley reflect the program
- what we want to do is apply the function to one entry and the use that accross the multiple projects

In [107]:
## How many projects have multiple funding componenets?
## a sample of those under the no_match flag with at least two entires in system

(no_match
 >>group_by(_.recipient_project_number, _.project_number, _.implementing_agency, _.project_title)
 >>summarize(n_descriptions = _.improvement_type_description.nunique())
 >>filter(_.n_descriptions>1)
 >>arrange(_.n_descriptions, _.n_descriptions)
)

recipient_project_number project_number implementing_agency  \
23               0113000123S        P020198          California   
27               0117000018S        P020199          California   
30               0213000002S        0059180          California   
31               0216000044S        0058644          California   
35               0300020624S        P020179          California   
..                       ...            ...                 ...   
136              0615000046S        0055155          California   
158              0716000064S        0053074          California   
192              0900000030S        P395348          California   
197              1013000266S        S108057          California   
229              1218000034S        X059084          California   

                                                                                                                                             project_title  \
23          IN MENDOCINO COUNTY NEAR UKIAH FROM 0.3 MILE WEST OF RUSSIAN RIVER BRIDGE AND OVERHEAD TO 0.4 MILE EAST OF COUNTY ROAD 144 REPLACE TWO BRIDGES   
27   IN LAKE COUNTY NEAR CLEAR LAKE OAKS AT BERYL WAY THIS PROJECT PROPOSES TO INSTALL A COMBINATION OF TWO ALTERNATIVE IN-LINE TERMINAL SYSTEMS, A RAIL E   
30   IN SISKIYOU COUNTY, AT VARIOUS LOCATIONS FROM 0.2 MILE NORTH OF REAM AVENUE OVERCROSSING TO 0.2 MILE SOUTH OF HILT ROAD OVERCROSSING. ENHANCE TRAFFIC   
31             IN TEHAMA COUNTY NEAR RED BLUFF FROM NINE MILE HILL OVERCROSSING TO 0.1 MILE NORTH OF BOWMAN ROAD OVERCROSSING. IMPROVE CLEAR RECOVERY ZONE   
35   NEAR LAKE WILDWOOD FROM 0.4 MILE EAST OF MCGANNERY LANE TO YUBA/NEVADA COUNTY LINE; ALSO IN NEVADA COUNTY, FROM YUBA/NEVADA COUNTY LINE TO MOONEY FLA   
..                                                                                                                                                     ...   
136   FRESNO COUNTY NEAR MENDOTA FROM 1.2 MILES NORTH OF HUDSON AVENUE OVERCROSSING TO 0.4 MILE SOUTH OF MANNING AVENUE OVERCROSSING. REPLACE  TUMEY GULCH   
158   NEAR CASTAIC, FROM LAKE HUGHES ROAD UNDERCROSSING TO 1.9 MILES NORTH OF TEMPLIN HIGHWAY. REPLACE APPROACH SLABS, UPGRADE DRAINAGE SYSTEMS, UPGRADING   
192  IN INYO COUNTY NEAR OLANCHA AND CARTAGO FROM 1.4 MILES SOUTH OF LOS ANGELES AQUEDUCT BRIDGE TO 0.2 MILE SOUTH OF ASH CREEK BRIDGE. CONSTRUCT 4 LANE E   
197  TUOLUMNE COUNTY IN AND NEAR LONG BARN FROM 0.1 MILE WEST OF LONG BARN ROAD TO 2.3 MILES EAST OF COW CREEK ROAD COLD PLANE, HMA, DRAINAGE REPLACEMENT,   
229  CONST BRIDGES, RW, SW, JPCP, HMA PVMT, DRAINAGE AND ELECT SYSTEMS. ORANGE COUNTY IN IRVINE, SANTA ANA, AND TUSTIN ON ROUTE 5 AT NEWPORT AVENUE UNDERC   

     n_descriptions  
23                2  
27                2  
30                2  
31                2  
35                2  
..              ...  
136               4  
158               4  
192               4  
197               4  
229               4  

[139 rows x 5 columns]

* of the 133 projects wiht unique project IDs, 60 have more than 1 entry
* using all unique regardless of numbers 

In [136]:
df_copy = (df>>filter(_.project_type==('Project')))

In [109]:
len(df>>filter(_.project_type==('Project'))
 >>group_by(_.fmis_transaction_date, _.program_code, _.program_code_description, _.project_number,
                                                   _.recipient_project_number, _.project_title, _.county_code, _.congressional_district,
                                                   _.summary_recipient_defined_text_field_1_value, _.implementing_agency_locode,
                                                   _.implementing_agency, _.district, _.county_name, _.rtpa_name, _.mpo_name)
 >>summarize(n_descriptions = _.improvement_type_description.nunique())>>filter(_.n_descriptions>1)>>arrange(_.n_descriptions, _.n_descriptions))

60

In [110]:
len(df>>filter(_.project_type==('Project'))
 >>group_by(_.fmis_transaction_date, _.program_code, _.program_code_description, _.project_number,
                                                   _.recipient_project_number, _.project_title, _.county_code, _.congressional_district,
                                                   _.summary_recipient_defined_text_field_1_value, _.implementing_agency_locode,
                                                   _.implementing_agency, _.district, _.county_name, _.rtpa_name, _.mpo_name)
 >>summarize(n_descriptions = _.improvement_type_description.nunique()))

133

In [146]:
proj_unique = (df>>filter(_.project_type==('Project'))
#  >>group_by(_.fmis_transaction_date, _.program_code, _.program_code_description, _.project_number,
#                                                    _.recipient_project_number, _.project_title, _.county_code, _.congressional_district,
#                                                    _.summary_recipient_defined_text_field_1_value, _.implementing_agency_locode,
#                                                    _.implementing_agency, _.district, _.county_name, _.rtpa_name, _.mpo_name)
# >>arrange(_.project_number, _.obligations_amount)
              )

#### Use .max() to get the max value for obligation amount (the main portion of the funding)

In [147]:
#proj_unique.drop_duplicates('project_number', keep='last')
proj_unique = proj_unique.groupby(['fmis_transaction_date','program_code','project_number','recipient_project_number','project_title',
                    'implementing_agency', 'district', 'county_name'])['obligations_amount'].max().reset_index()

In [148]:
merge_cols = ['fmis_transaction_date', 'program_code','project_number','recipient_project_number','project_title',
                    'implementing_agency', 'county_name', 'district', 'obligations_amount']

#### Merge to get full details

In [149]:
proj_unique = (pd.merge(proj_unique, df_copy, how='left', on=merge_cols))

In [151]:
proj_unique.sample(4)

fmis_transaction_date program_code project_number  \
32             2022-04-26         Y001        NBIS522   
204            2022-09-12         YS30        P204006   
73             2022-05-16         Y001        P057069   
193            2022-09-09         Y240        5177047   

    recipient_project_number  \
32              0000001453L5   
204              0617000297S   
73               0714000067S   
193              0422000384L   

                                                                                                                                             project_title  \
32                                                     STATEWIDE - IN SERVICE BRIDGES OWNED BY LOCAL AGENCIES FEDERALLY MANDATED BRIDGE INSPECTION PROGRAM   
204                                                          KERN COUNTY IN BAKERSFIELD AT 8TH STREET. INSTALLATION OF HIGH INTENSITY ACTIVATED CROSSWALK.   
73   LA-057, IN LOS ANGELES COUNTY IN DIAMOND BAR AND CITY OF INDUSTRY FROM ORANGE COUNTY LINE TO ROUTE 57/60 SEPARATION THE PROJECT IS LOCATED IN LOS ANG   
193  OYSTER POINT BLVD BETWEEN VETERANS BLVD AND ECCLES AVE, GATEWAY BLVD BETWEEN OYSTER POINT BLVD AND E GRAND AVE, AND EGRAND AVE BETWEEN GATEWAY BLVD A   

     implementing_agency  district        county_name  obligations_amount  \
32            California      3.00  Sacramento County       14,164,800.00   
204           California      6.00        Kern County        1,170,500.00   
73            California     12.00      Orange County           23,276.43   
193  South San Francisco      4.00   San Mateo County           46,411.00   

           program_code_description  county_code congressional_district  \
32       NATIONAL HIGHWAY PERF IIJA        67.00            Cong Dist 3   
204    HIGHWAY SAFETY IMP PROG IIJA        29.00           Cong Dist 21   
73       NATIONAL HIGHWAY PERF IIJA        59.00           Cong Dist 39   
193  SURFAC TRNSP BLK GRTS-FLX IIJA        81.00           Cong Dist 14   

     improvement_type                   improvement_type_description  \
32              49.00  Bridge Inspection and Bridge Related Training   
204             21.00                                         Safety   
73              17.00                       Construction Engineering   
193             17.00                       Construction Engineering   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
32                                  S    NON-MPO                         NaN   
204                                    S    KCOG                         NaN   
73                                     S    SCAG                         NaN   
193                                     L5177MTC                    5,177.00   

                                  rtpa_name  \
32                                      NaN   
204                                     NaN   
73                                      NaN   
193  Metropolitan Transportation Commission   

                                   mpo_name project_method project_type  \
32                                      NaN                     Project   
204                                     NaN        Install      Project   
73                                      NaN                     Project   
193  Metropolitan Transportation Commission                     Project   

                    project_name_new  
32      Project in Sacramento County  
204   Install Project in Kern County  
73          Project in Orange County  
193   Project in South San Francisco

#### run with update_no_matched function

In [160]:
proj_unique_new = update_no_matched(proj_unique, 'project_type', 'improvement_type_description', 'implementing_agency')

In [162]:
proj_unique_new.sample(5)

fmis_transaction_date program_code project_number  \
51             2022-05-05         Y233        5946138   
16             2022-03-07         Y301        5108202   
140            2022-07-25         YS30        5921080   
194            2022-09-09         YS30        5051005   
63             2022-05-10         Y001        Q101396   

    recipient_project_number  \
51               0614000078L   
16               0722000055L   
140              0419000574L   
194              0420000045L   
63               0517000001S   

                                                                                                                                             project_title  \
51                                           BRIDGE 46C0004 ON COUNTY ROAD D112 OVER NORTH BRANCH TULE RIVER REPLACE 2 LANE BRIDGE WITH 2 LANE BRIDGE (TC)   
16   DOWNTOWN LONG BEACH (BOUNDED BY MAGNOLIA AVE TO THE WEST, OCEAN BLVD TO THE SOUTH, 7TH STREET TO THE NORTH, AND ALAMITOS AVE TO THE EAST). NON-INFRAS   
140  SIVERADO TRAIL: NORTH OF TRANCAS ST. AND SOUTH OF SR 29; EXCLUDING PORTIONS WITHIN JURISDICTION OF CITY OF ST. HELENA, CITY OF CALISTOGA, AND CALTRAN   
194   ON SILVERADO TRAIL, WITH THE PROJECT LIMITS BORDERING NAPA COUNTY, BEGINNING FROM APPROXIMATELY 175 FEET NORTH OF HOWELL MOUNTAIN ROAD AND ENDING AT   
63    SANTA BARBARA AND SAN LUIS OBISPO COUNTIES AT VARIOUS LOCATIONS FROM ROUTE 135 TO WEST TEFFT STREET CONSTRUCT CONTRASTING SURFACE TREATMENT AND GORE   

    implementing_agency  district             county_name  obligations_amount  \
51        Tulare County      6.00           Tulare County        1,391,304.35   
16           Long Beach      7.00      Los Angeles County          543,000.00   
140         Napa County      4.00             Napa County        1,470,217.08   
194           St Helena      4.00             Napa County           66,500.00   
63           California      5.00  San Luis Obispo County          843,287.44   

          program_code_description  county_code congressional_district  \
51     STBG IIJA OFF-SYSTEM BRIDGE       107.00           Cong Dist 22   
16   TRANS ALTERNATIVES >200K IIJA        37.00           Cong Dist 47   
140   HIGHWAY SAFETY IMP PROG IIJA        55.00            Cong Dist 5   
194   HIGHWAY SAFETY IMP PROG IIJA        41.00            Cong Dist 2   
63      NATIONAL HIGHWAY PERF IIJA        79.00           Cong Dist 24   

     improvement_type            improvement_type_description  \
51              11.00  Bridge Replacement - No Added Capacity   
16              44.00                                   Other   
140             21.00                                  Safety   
194             21.00                                  Safety   
63               6.00       4R - Restoration & Rehabilitation   

    summary_recipient_defined_text_field_1_value  implementing_agency_locode  \
51                                     L5946TCAG                    5,946.00   
16                                     L5108SCAG                    5,108.00   
140                                     L5921MTC                    5,921.00   
194                                     L5051MTC                    5,051.00   
63                                    S    SBCAG                         NaN   

                                                rtpa_name  \
51               Tulare County Association of Governments   
16   Los Angeles County Metropolitan Transportation Auth.   
140                Metropolitan Transportation Commission   
194                Metropolitan Transportation Commission   
63                                                    NaN   

                                           mpo_name project_method  \
51         Tulare County Association Of Governments        Replace   
16   Southern California Association Of Governments                  
140          Metropolitan Transportation Commission                  
194          Metropolitan Transportation Commission                 

In [164]:
len(proj_unique_new>>filter(_.project_name_new2==""))

24

In [165]:
(proj_unique_new>>filter(_.project_name_new2=="")>>select(_.project_number, _.recipient_project_number, _.project_title, _.improvement_type_description, _.implementing_agency, _.county_name,
                                                            _.project_method, _.project_type, _.project_name_new, _.project_name_new2))

project_number recipient_project_number  \
2          6480013              0417000451L   
11         5917084              0314000112L   
16         5108202              0722000055L   
20         6084277              0422000334L   
32         NBIS522             0000001453L5   
56         P001676             0716000335S9   
82         6084279              0422000413L   
96         0052029              1216000044S   
103        4053001              0716000031S   
104        P014105              0716000029S   
121        5065024              0812000141L   
122        5946142              0614000082L   
133        5937206              0417000076L   
134        5937207              0417000077L   
135        6236020              0322000257L   
136        5937205              0416000438L   
137        6085086              0322000305L   
153        6084288              0423000020L   
160        P070106              0312000068S   
161        6211149              1119000022L   
200        P905023              1120000088S   
213        NBIL539              0815000215L   
215        5006664              0700020306L   
226        P051034              0312000054S   

                                                                                                                                             project_title  \
2    COUNTY WIDE- APPROXIMATELY 300 PUBLIC SCHOOL ALAMEDA COUNTY SAFE ROUTS TO SCHOOL PROGRAM OUTREACH AND EDUCATION. WORK PROGRAM IS FOCUSED ON SUPPORTIN   
11           DOG BAR ROAD AT BEAR RIVER ~0.5 MILES SOUTH EAST OF MAGNOLIA ROAD AT COUNTY LINE. BR. # 17C0031 REPLACE 1 LANE BRIDGE WITH 2 LANE BRIDGE (TC)   
16   DOWNTOWN LONG BEACH (BOUNDED BY MAGNOLIA AVE TO THE WEST, OCEAN BLVD TO THE SOUTH, 7TH STREET TO THE NORTH, AND ALAMITOS AVE TO THE EAST). NON-INFRAS   
20   REGIONWIDE - NINE COUNTY SAN FRANCISCO BAY AREA THE REGIONAL MAPPING DATA SERVICES SYSTEM WILL EXPOSE A SET OF PREDEFINED MAPPING-RELATED DATA APIS.(   
32                                                     STATEWIDE - IN SERVICE BRIDGES OWNED BY LOCAL AGENCIES FEDERALLY MANDATED BRIDGE INSPECTION PROGRAM   
56    PROJECT ID 0716000335   (EA 07-4X970)  LA-001-42.5/42.5   (PPNO-5059)ROUTE 001: IN MALIBU, SOUTH OF BIG ROCK DRIVE. SHORELINE EMBANKMENT RESTORATION   
82                                                                             SF BAY AREA: REGIONWIDE TECHNICAL ASSISTANCE MOBILITY HUB PILOT PROGRAM(TC)   
96   ORANGE COUNTY, IN ANAHEIM FROM SANTA ANA STREET OVERCROSSING TO 0.5 MILE NORTH OF LA PALMA AVENUE OVERCROSSING PLANTING, IRRIGATION, ROCK BLANKET, MI   
103  IN LOS ANGELES COUNTY IN LOS ANGELES FROM 0.2 MILE SOUTH OF BURBANK BOULEVARD OVERCROSSING TO 0.3 MILE NORTH OF RINALDI STREET UNDERCROSSING PLANT VE   
104   IN AND NEAR SANTA CLARITA, FROM CANYON PARK BOULEVARD TO SPRING CANYON ROAD UNDERCROSSING. MINOR CONCRETE BEYOND GORE, SLOPE PAVE BENEATH ABUTMENTS,   
121  BARTON ROAD OVERHEAD AT UNION PACIFIC RAILROAD, 0.25 MILE WEST OF I-215, BR. NO. 54C-0379 REPLACE OVERHEAD WITH TWO-LANE ROAD ACROSS DISCONTINUED RAI   
122                                                             BRIDGE 46C0340 ON AVENUE 428 OVER SAND CREEK REPLACE 2 LANE BRIDGE WITH 2 LANE BRIDGE (TC)   
133  ALDERCROFT HEIGHTS ROAD OVER LOS GATOS CREEK, 2.12 MILES S OF LOS GATOS CREEL ROAD. BRIDGE 37C0507 REPLACE EXISTING ONE LANE BRIDGE WITH A NEW 2 LANE   
134  ALDERCROFT HEIGHTS ROAD OVER LOS GATOS CREEK 2.38 MILES SOUTH OF LOS GATOS SC ROAD. BRIDGE 37C0508 REPLACE EXISTING ONE-LANE BRIDGE WITH A NEW 2-LANE   
135                                                      IN SACRAMENTO COUNTY. PHASE 3 SACRAMENTO EMERGENCY CLEAN AIR TRANSPORTATION PROGRAM (SECAT). (TC)   
136  ALDER CROFT HEIGHTS ROAD OVER HOOKER CREEK, 2.12 MILES SOUTH OF LOS GATOS SC ROAD.   BRIDGE NO. 37C0506 REPLACE EXISTING ONE LANE BRIDGE WITH A NEW T   
137  SACOG REGION - IN SACRAMENTO, SUTTER, YOLO AND YUBA COUNTIES CONDUCT NECESSARY PLANNING, STATE AND FEDERAL PROGRAMMING, AND MONITORING AC